internal/pipeline/pipeline.go

```diff
@@ -164,3 +196,90 @@ type Function struct {
 }
+
+// Validate will validate all fields in function.
+func (f *Function) Validate() error {
+	err := ValidateFunctionName(f.Image)
+	if err != nil {
+		return fmt.Errorf("'image' is invalid: %w", err)
+	}
+
+	var configFields []string
+	if f.ConfigPath != "" {
+		if err := ValidatePath(f.ConfigPath); err != nil {
+			return fmt.Errorf("'configPath' is invalid: %w", err)
+		}
+		configFields = append(configFields, "configPath")
+	}
+	if len(f.ConfigMap) != 0 {
+		configFields = append(configFields, "configMap")
+	}
+	if !isNodeZero(&f.Config) {
+		configFields = append(configFields, "config")
+	}
+	if len(configFields) > 1 {
+		return fmt.Errorf("following fields are mutually exclusive: 'config', 'configMap', 'configPath'. Got %q",
+			strings.Join(configFields, ", "))
+	}
+
+	return nil
+}
+
+// ValidateFunctionName validates the function name.
+// According to Docker implementation
+// https://github.com/docker/distribution/blob/master/reference/reference.go. A valid
+// name definition is:
+//	name                            := [domain '/'] path-component ['/' path-component]*
+//	domain                          := domain-component ['.' domain-component]* [':' port-number]
+//	domain-component                := /([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9])/
+//	port-number                     := /[0-9]+/
+//	path-component                  := alpha-numeric [separator alpha-numeric]*
+// 	alpha-numeric                   := /[a-z0-9]+/
+//	separator                       := /[_.]|__|[-]*/
+func ValidateFunctionName(name string) error {
+	pathComponentRegexp := `(?:[a-z0-9](?:(?:[_.]|__|[-]*)[a-z0-9]+)*)`
+	domainComponentRegexp := `(?:[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9])`
+	domainRegexp := fmt.Sprintf(`%s(?:\.%s)*(?:\:[0-9]+)?`, domainComponentRegexp, domainComponentRegexp)
+	nameRegexp := fmt.Sprintf(`^(?:%s\/)?%s(?:\/%s)*$`, domainRegexp,
+		pathComponentRegexp, pathComponentRegexp)
+	matched, err := regexp.MatchString(nameRegexp, name)
+	if err != nil {
+		return err
+	}
+	if !matched {
+		return fmt.Errorf("function name %q is invalid", name)
+	}
+	return nil
+}
+
+// IsNodeZero returns true if all the public fields in the Node are empty.
+// Which means it's not initialized and should be omitted when marshal.
+// The Node itself has a method IsZero but it is not released
+// in yaml.v3. https://pkg.go.dev/gopkg.in/yaml.v3#Node.IsZero
+// TODO: Use `IsYNodeZero` method from kyaml when kyaml has been updated to
+// >= 0.10.5
+func isNodeZero(n *yaml.Node) bool {
+	return n != nil && n.Kind == 0 && n.Style == 0 && n.Tag == "" && n.Value == "" &&
+		n.Anchor == "" && n.Alias == nil && n.Content == nil &&
+		n.HeadComment == "" && n.LineComment == "" && n.FootComment == "" &&
+		n.Line == 0 && n.Column == 0
+}
+
+// ValidatePath validates input path and return an error if it's invalid
+func ValidatePath(p string) error {
+	if path.IsAbs(p) {
+		return fmt.Errorf("path %q is not relative", p)
```

Comment:
In general, you can remove some duplication, the caller can wrap errors returned by this function and add additional context.

return fmt.Errorf("path is not relative")

And then caller of ValidatePath can wrap the error:

sources[0] is invalid: '/a/b/c'. path is relative